File: solutions/Golang/binary_tree_zigzag_level_order_traversal_parity.go
package main

type TreeNode struct { Val int; Left, Right *TreeNode }

func ZigzagLevelOrder(root *TreeNode) [][]int {
    var result [][]int
    var dfs func(*TreeNode, int)
    dfs = func(n *TreeNode, level int) {
        if n == nil { return }
        if level == len(result) { result = append(result, []int{}) }
        if level%2 == 0 { result[level] = append(result[level], n.Val) }
        else { result[level] = append([]int{n.Val}, result[level]...) }
        dfs(n.Left, level + 1)
        dfs(n.Right, level + 1)
    }
    dfs(root, 0)
    return result
}

File: solutions/Java/binary_tree_zigzag_level_order_traversal_parity.java
import java.util.*;

public class Solution {
    void dfs(TreeNode n, int level, List<List<Integer>> result) {
        if (n == null) return;
        if (level == result.size()) result.add(new ArrayList<>());
        if (level % 2 == 0) result.get(level).add(n.val);
        else result.get(level).add(0, n.val);
        dfs(n.left, level + 1, result);
        dfs(n.right, level + 1, result);
    }
    public List<List<Integer>> zigzagLevelOrder(TreeNode root) {
        List<List<Integer>> result = new ArrayList<>();
        dfs(root, 0, result);
        return result;
    }
    public static class TreeNode { int val; TreeNode left, right; TreeNode(int x) { val = x; } }
}
